javascript/ql/src/semmle/javascript/StringOps.qll

```diff
@@ -492,15 +537,220 @@ module StringOps {
     }
+
+    /**
+     * Gets the enclosing concatenation in which this is an operand, if any.
+     */
+    pragma[inline]
+    Concatenation getParentConcatenation() {
+      this = StringConcatenation::getAnOperand(result)
+    }
+
+    /**
+     * Gets the last leaf in this concatenation tree.
+     *
+     * For example, `z` is the last leaf in `x + y + z`.
+     */
+    pragma[inline]
+    ConcatenationLeaf getLastLeaf() {
+      result = StringConcatenation::getLastOperand*(this)
+    }
+
+    /**
+     * Gets the first leaf in this concatenation tree.
+     *
+     * For example, `x` is the first leaf in `x + y + z`.
+     */
+    pragma[inline]
+    ConcatenationLeaf getFirstLeaf() {
+      result = StringConcatenation::getFirstOperand*(this)
+    }
+
+    /**
+     * Gets the leaf that is occurs immediately before this leaf in the
+     * concatenation tree, if any.
+     *
+     * For example, `y` is the previous leaf from `z` in `x + y + z`.
+     */
+    pragma[inline]
+    ConcatenationLeaf getPreviousLeaf() {
+      adjacentLeaves(result, this)
+    }
+
+    /**
+     * Gets the leaf that is occurs immediately after this leaf in the
+     * concatenation tree, if any.
+     *
+     * For example, `y` is the next leaf from `x` in `x + y + z`.
+     */
+    pragma[inline]
+    ConcatenationLeaf getNextLeaf() {
+      adjacentLeaves(this, result)
+    }
+  }
+
+  /**
+   * A data flow node that performs a string concatenation and returns the result.
+   *
+   * Examples:
+   * ```
+   * x + y
+   * x += y
+   * [x, y].join('')
+   * Array(x, y).join('')
+   * `Hello ${message}`
+   * ```
+   *
+   * See `ConcatenationNode` for more information.
+   */
+  class Concatenation extends ConcatenationNode {
+    pragma[inline]
+    Concatenation() {
+      exists(StringConcatenation::getAnOperand(this))
+    }
+  }
+
+  /**
+   * One of the operands in a string concatenation.
+   *
+   * Examples:
+   * ```
+   * x + y              // x and y are operands
+   * [x, y].join('')    // x and y are operands
+   * `Hello ${message}` // `Hello ` and message are operands
+   * ```
+   *
+   * See `ConcatenationNode` for more information.
+   */
+  class ConcatenationOperand extends ConcatenationNode {
+    pragma[inline]
+    ConcatenationOperand() {
+      this = StringConcatenation::getAnOperand(_)
+    }
+  }
+
+  /**
+   * A data flow node that performs a string concatenation, and is not an
+   * immediate operand in a larger string concatenation.
+   *
+   * Examples:
+   * ```
+   * // x + y + z is a root, but the inner x + y is not
+   * return x + y + z;
+   * ```
+   *
+   * See `ConcatenationNode` for more information.
+   */
+  class ConcatenationRoot extends Concatenation {
+    pragma[inline]
+    ConcatenationRoot() {
+      isRoot()
+    }
+
+    /**
+     * Gets a leaf in this concatenation tree that this node is the root of.
+     */
+    pragma[inline]
+    ConcatenationOperand getALeaf() {
```

Comment:
Should the result type be `ConcatenationLeaf`?

Reply:
Yikes, good catch. It most certainly should be.